### DISCUSSION TITLE
use client but still complains about useLayoutEffect

### DISCUSSION THREAD

**Discussion by thanosoncode** [vercel/next.js#49992] [App Router]
I have marked a component as client using the "use client" on the top of the file, nonetheless I get warning that "useLayoutEffect should only be used in components that render exclusively on the client"

**Comment by devjiwonchoi** [comment]
@thanosoncode Could you share your component code?

**Comment by icyJoseph** [comment]
Hi,

The `use client` directive has some nuances. In particular, in Next.js pages can be pre-rendered, ever since the pages directory. That means that, even if a component uses the `use client` directive, it'll be pre-rendered at build time, and that happens within a sever side context, not only that, if the component is part of a server side rendering pass, on demand, it'll also be rendered server side, unless it suspends.

Think for a moment, in terms of SEO, or users who don't have JS enabled, as well as bots, crawlers, etc, those would miss on the HTML of `use client` components.

![Screenshot 2023-05-18 at 16 25 16](https://github.com/vercel/next.js/assets/21013447/adb57c16-f63a-4fe6-9d98-395468231574)

> Components in the Client Component module graph are primarily rendered on the client, but with Next.js, they can also be pre-rendered on the server and hydrated on the client.

The thing with `useLayoutEffect` is that it runs before the browser paints, and in the server that's a bit of a problem... mostly because `useLayoutEffect` won't run, but there might be an expectation that it does, but also there's no DOM to interact with, because we are server side, which is what you typically, want to do with `useLayoutEffect`.

The warning is mostly referring to this gist: https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85, where a component is rendered by using a `useState + useEffect` combo:

- using a piece of React state, either render or don't a given component
- start with state set to false
- inside a useEffect, set the state to true - because useEffect runs only client side
- now you render the component, which in your case, uses `useLayoutEffect`

Last but not least, some people, understanding the consequences, use the https://usehooks-ts.com/react-hook/use-isomorphic-layout-effect hook :)

**Reply by devjiwonchoi** [reply]
I was writing the same answer but am astonished at how descriptive and organized the amount of information is. I want to be like you one day.
